deadline_/social/models.py

```diff
@@ -0,0 +1,384 @@
+                            content={
+                                'submission_id': comment.submission.id,
+                                'challenge_id': comment.submission.challenge.id,
+                                'challenge_name': comment.submission.challenge.name,
+                                'commenter_name': comment.author.username,
+                                'commenter_id': comment.author.id,
+                                'comment_content': comment.content,
+                                'comment_id': comment.id
+                            })
+
+    def create_submission_comment_reply_notification(self, comment: SubmissionComment):
+        if comment.parent.author == comment.author:
+            return
+
+        return self._create(recipient=comment.parent.author, type=RECEIVE_SUBMISSION_COMMENT_REPLY_NOTIFICATION,
+                           content={
+                               'submission_id': comment.submission.id,
+                               'challenge_id': comment.submission.challenge.id,
+                               'challenge_name': comment.submission.challenge.name,
+                               'commenter_name': comment.author.username,
+                               'commenter_id': comment.author.id,
+                               'comment_content': comment.content,
+                               'comment_id': comment.id
+                           })
+
+    def create_challenge_comment_reply_notification(self, reply: ChallengeComment):
+        if reply.parent.author == reply.author:
+            return
+
+        return self._create(recipient=reply.parent.author, type=RECEIVE_CHALLENGE_COMMENT_REPLY_NOTIFICATION,
+                            content={
+                                'challenge_id': reply.challenge.id,
+                                'challenge_name': reply.challenge.name,
+                                'comment_id': reply.id,
+                                'comment_content': reply.content,
+                                'commenter_id': reply.author.id,
+                                'commenter_name': reply.author.username
+                            })
+
+
+class Notification(models.Model):
+    """
+    A Notification is a simple notification that a user receives, again facebook-esque
+    Since the content here is dynamic (somebody liked your post, new challenge appears in the site, etc),
+        we need an HStore field to store data related to the type of the notification.
+    """
+    recipient = models.ForeignKey(User)
+    type = models.CharField(max_length=60)  # no other table for now
+    content = HStoreField()  # varies depending on the type
+    is_read = models.BooleanField(default=False)
+    created_at = models.DateTimeField(auto_now_add=True)
+    updated_at = models.DateTimeField(auto_now=True)
+
+    objects = NotificationManager()
+    # TODO: Implement some logic to squash multiple notifications into one
```

Comment:
I think this is nice to have. :)

Reply:
I'm not sure if it's right to have such logic.
Regardless, __I am implementing it now__. 
But I want to note my concerns: I wonder if it's a bad idea to essentially 'erase' Notifications by merging them into one. 
My implementation would not save a new notification at all but rather update the previous one. e.g User 1 liked Post -> Notification created, User 2 liked Post -> old Notification updated to say "Two users liked your post".
Reminder: Only certain type of notifications can be squashed as some do not make sense to be squashed into one

There is a possibility for a race condition to occur -> The front-end queries for the old notification and gets it. At that very moment, a new like occurs and the back-end updates the old notification. The front-end will work with the old notification and in reality will never query for it again. This way the 'new' notification gets lost essentially. I want to note that this isn't a big deal and is something we can easily ignore imo.

I just wanted to share that an alternative approach would be to not do anything in the back-end and continue sending down notifications - and have the front-end build them into one. This makes more semantic and logical sense, as potentially in different front-ends you might want to handle multiple notifications differently? But this will be too much work to have the front-end group notifications. It's also a useless **over-complication** of something that doesn't make practical sense. That's why I'm sticking with the back-end doing the work.

Reply:
Let's discuss this on the next meeting.